src/main/java/com/google/api/gax/grpc/ServiceApiSettings.java

```diff
@@ -66,34 +98,51 @@
   /**
    * Constructs an instance of ServiceApiSettings.
    */
-  protected ServiceApiSettings(ManagedChannel channel,
-                               boolean shouldAutoCloseChannel,
-                               ScheduledExecutorService executor,
-                               ConnectionSettings connectionSettings,
-                               String generatorName,
-                               String generatorVersion,
-                               String clientLibName,
-                               String clientLibVersion) {
-    this.channel = channel;
-    this.executor = executor;
-    this.connectionSettings = connectionSettings;
-    this.shouldAutoCloseChannel = shouldAutoCloseChannel;
+  protected ServiceApiSettings(
+      ChannelProvider channelProvider,
+      ExecutorProvider executorProvider,
+      String generatorName,
+      String generatorVersion,
+      String clientLibName,
+      String clientLibVersion) {
+    this.channelProvider = channelProvider;
+    this.executorProvider = executorProvider;
     this.clientLibName = clientLibName;
     this.clientLibVersion = clientLibVersion;
     this.generatorName = generatorName;
     this.generatorVersion = generatorVersion;
   }
 
-  public final ManagedChannel getChannel() {
-    return channel;
+  /**
+   * Return the channel to be used to connect to the service, retrieved using the channelProvider.
+   * If no channel was set, a default channel will be instantiated.
+   */
+  public final ManagedChannel getOrBuildChannel() throws IOException {
+    return getChannelProvider().getChannel(getOrBuildExecutor());
```

Comment:
Channel re-use is ok when a fixed channel is passed in and shouldAutoCloseChannel is false, but not ok when shouldAutoCloseChannel is true. We might want to enforce that in some way. 

I realized we have a resource leak with executors - our XApi class will close a channel if the default one is created, but not an executor if the default one is created. We should probably manage its lifetime in the same way. Although, we have a slightly tricksy situation if auto-close is true on an executor and false on a channel; the channel will no longer work on top of the executor that was closed. We should probably disallow the broken combination there.